Zend/zend_builtin_functions.stub.php

```diff
@@ -68,24 +77,46 @@ function function_exists(string $function): bool {}
 
 function class_alias(string $class, string $alias, bool $autoload = true): bool {}
 
+/**
+ * @return array<int, string>
+ * @refcount 1
+ */
 function get_included_files(): array {}
 
-/** @alias get_included_files */
+/**
+ * @return array<int, string>
+ * @refcount 1
+ * @alias get_included_files
+ */
 function get_required_files(): array {}
 
 function trigger_error(string $message, int $error_level = E_USER_NOTICE): bool {}
 
 /** @alias trigger_error */
 function user_error(string $message, int $error_level = E_USER_NOTICE): bool {}
 
-/** @return string|array|object|null */
+/**
+ * @return string|array<int, string|object>|object|null
+ * @refcount N
+ */
 function set_error_handler(?callable $callback, int $error_levels = E_ALL) {}
 
+/**
+ * @return true
+ * @refcount 0
```

Comment:
You are currently only generating func info if `@refcount` is specified. Would it make sense to generate it if there is a more specific doc type without `@refcount`? The refcount would then default to 0 (scalars) or N (other). As is, this `@refcount 0` info is redundant, because bool cannot be anything else.

Reply:
Yes, this is also something I wasn't very sure about. The reason why I opted for requiring refcount is that determining if the doc return type is more specific than the native one is not easy. If I'm not mistaken, basically I should implement variance checks. Of course, we could also go simple by only checking for `@return true` and `@return array<>`/`@return array[]`. Is this ok?

Another reason of my choice is that this way we can explicitly control whether a function entry should be added to `zend_func_infos.h` or not. I'm not sure if this is a problem in practice, but in theory, a function which is not very commonly used could have additional return type hint in docblock.

Reply:
If a doc type is present, shouldn't it always be more specific than the proper type? Otherwise, the doc type should have been omitted.

Reply:
> If a doc type is present, shouldn't it always be more specific than the proper type? Otherwise, the doc type should have been omitted.

Hmm, you are right, that's true.

Then you'd prefer using the implicit refcount-based approach that you described above? I can also add a check somewhere to ensure that the return type hint != return type declaration

Reply:
For reference: `zend_func_info.c` currently contains 17 `FN()` and 22 `F0()` function entries, so I'd say that requiring `@refcount` to be defined wouldn't cause very much redundancy.

Reply:
I've just implemented the automatic detection of refcount in case of scalar return types + improved a few other related things (e.g. the validation of the native and phpdoc return types so that they are not the same)